File: java/739.daily-temperatures.java
/*
 * @lc app=leetcode id=739 lang=java
 *
 * [739] Daily Temperatures
 */

// @lc code=start
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        Deque<Integer> idx = new ArrayDeque();

        int[] res = new int[temperatures.length];

        for(int i = 0; i < temperatures.length; i++) {
            while(!idx.isEmpty() && temperatures[i] > temperatures[idx.peek()]) {
                int j = idx.pop(); 
                res[j] = i - j;
            }

            idx.push(i);
        }

        return res;
    }
}
// @lc code=end

File: cpp/739.daily-temperatures.cpp
/*
 * @lc app=leetcode id=739 lang=cpp
 *
 * [739] Daily Temperatures
 */

// @lc code=start
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        std::stack<int> stk;

        std::vector<int> res(temperatures.size(), 0);

        for(int i = 0; i < temperatures.size(); i++) {
            while(!stk.empty() && temperatures[i] > temperatures[stk.top()]) {
                res[stk.top()] = i - stk.top();
                stk.pop();
            }

            stk.push(i);
        }

        return res;
    }
};
// @lc code=end

